src/theory/arith/arith_msum.cpp

```diff
@@ -0,0 +1,371 @@
+  }
+}
+
+int ArithMSum::isolate(
+    Node v, const std::map<Node, Node>& msum, Node& veq_c, Node& val, Kind k)
+{
+  std::map<Node, Node>::const_iterator itv = msum.find(v);
+  if (itv != msum.end())
+  {
+    std::vector<Node> children;
+    Rational r =
+        itv->second.isNull() ? Rational(1) : itv->second.getConst<Rational>();
+    if (r.sgn() != 0)
+    {
+      for (std::map<Node, Node>::const_iterator it = msum.begin();
+           it != msum.end();
+           ++it)
+      {
+        if (it->first != v)
+        {
+          Node m;
+          if (!it->first.isNull())
+          {
+            if (!it->second.isNull())
+            {
+              m = NodeManager::currentNM()->mkNode(MULT, it->second, it->first);
+            }
+            else
+            {
+              m = it->first;
+            }
+          }
+          else
+          {
+            m = it->second;
+          }
+          children.push_back(m);
+        }
+      }
+      val = children.size() > 1
+                ? NodeManager::currentNM()->mkNode(PLUS, children)
+                : (children.size() == 1
+                       ? children[0]
+                       : NodeManager::currentNM()->mkConst(Rational(0)));
+      if (!r.isOne() && !r.isNegativeOne())
+      {
+        if (v.getType().isInteger())
+        {
+          veq_c = NodeManager::currentNM()->mkConst(r.abs());
+        }
+        else
+        {
+          val = NodeManager::currentNM()->mkNode(
+              MULT,
+              val,
+              NodeManager::currentNM()->mkConst(Rational(1) / r.abs()));
+        }
+      }
+      if (r.sgn() == 1)
+      {
+        val = negate(val);
+      }
+      else
+      {
+        val = Rewriter::rewrite(val);
+      }
+      return (r.sgn() == 1 || k == EQUAL) ? 1 : -1;
+    }
+  }
+  return 0;
+}
+
+int ArithMSum::isolate(
+    Node v, const std::map<Node, Node>& msum, Node& veq, Kind k, bool doCoeff)
+{
+  Node veq_c;
+  Node val;
+  // isolate v in the (in)equality
+  int ires = isolate(v, msum, veq_c, val, k);
+  if (ires != 0)
+  {
+    Node vc = v;
+    if (!veq_c.isNull())
+    {
+      if (doCoeff)
+      {
+        vc = NodeManager::currentNM()->mkNode(MULT, veq_c, vc);
+      }
+      else
+      {
+        return 0;
+      }
+    }
+    bool inOrder = ires == 1;
+    veq = NodeManager::currentNM()->mkNode(
+        k, inOrder ? vc : val, inOrder ? val : vc);
+  }
+  return ires;
+}
+
+Node ArithMSum::solveEqualityFor(Node lit, Node v)
+{
+  Assert(lit.getKind() == EQUAL);
+  // first look directly at sides
+  TypeNode tn = lit[0].getType();
+  for (unsigned r = 0; r < 2; r++)
+  {
+    if (lit[r] == v)
+    {
+      return lit[1 - r];
+    }
+  }
+  if (tn.isReal())
+  {
+    std::map<Node, Node> msum;
+    if (ArithMSum::getMonomialSumLit(lit, msum))
+    {
+      Node val, veqc;
+      if (ArithMSum::isolate(v, msum, veqc, val, EQUAL) != 0)
+      {
+        if (veqc.isNull())
```

Comment:
Merge these two if conditions?

Reply:
I don't like merging conditions where the order becomes critical (veqc is computed by the first condition)